onnxruntime/core/framework/bfc_arena.cc

```diff
@@ -411,6 +432,70 @@ void BFCArena::Free(void* p) {
   }
 }
 
+Status BFCArena::Shrink() {
+  std::lock_guard<OrtMutex> lock(lock_);
+  auto num_regions = region_manager_.regions().size();
+  std::vector<void*> region_ptrs;
+  std::vector<size_t> region_sizes;
+  region_ptrs.reserve(num_regions);
+  region_sizes.reserve(num_regions);
+
+  for (const auto& region : region_manager_.regions()) {
+    if (consider_first_allocation_region_for_shrinkage_ || region.id() != 0) {
+      region_ptrs.push_back(region.ptr());
+      region_sizes.push_back(region.memory_size());
+    }
+  }
+
+  size_t i = 0;
+  for (void* region_ptr : region_ptrs) {
+    bool deallocate_region = true;
+    ChunkHandle region_begin_chunk = region_manager_.get_handle(region_ptr);
+    ChunkHandle h = region_begin_chunk;
+    while (h != kInvalidChunkHandle) {
+      const Chunk* c = ChunkFromHandle(h);
+      if (c->in_use()) {
+        // at-least one used chunk found in the allocation region -
+        // so we cannot deallocate it
+        deallocate_region = false;
+        break;
+      }
+      h = c->next;
+    }
+
+    if (deallocate_region) {
+      auto shrink_size = region_sizes[i];
+      stats_.num_arena_shrinkages += 1;
+      stats_.total_allocated_bytes -= shrink_size;
+
+      LOGS_DEFAULT(VERBOSE) << device_allocator_->Info().name << " BFC Arena shrunk by "
```

Comment:
There is a nit to be ironed out here. The reported memory allocated by the arena here is total_allocated_bytes which is allocation by extension + allocation by reserve.

If someone has chosen to allocate initializers using Reserve() and choose the shrinkage option, let us say the initializers take up 175MB. The Run call results in 2 extensions of size: 1MB and 2MB and once the 2MB allocation is shrunk (the 1MB will not be shrunk as it is the first allocation), we will see the stat reported: the total allocated bytes is 175MB + 1MB even after the shrink (people might expect to see just 1MB).

Reply:
I think it may not be that bad - total_allocated_bytes_ does need to include allocations from Reserve() calls logically. We needn't expose a stat for allocations through Extend() alone - it seems like overkill.